File: lcode/ex10/generateParenthesis.go
// Copyright 2019 readailib.com. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package main

import "fmt"

// 给出 n 代表生成括号的对数，请你写出一个函数，
// 使其能够生成所有可能的并且有效的括号组合。
// n = 3
// [
//   "((()))",
//   "(()())",
//   "(())()",
//   "()(())",
//   "()()()"
// ]
var res []string

func generateParenthesis(n int) []string {
	res = make([]string, 0)
	walk("", 0, 0, n)
	return res
}

func walk(str string, left, right, n int) {

	if left < right || left > n || right > n {
		return
	}

	if left == n && right == n {
		res = append(res, str)
		return
	}

	walk(str+"(", left+1, right, n)
	walk(str+")", left, right+1, n)
	return
}

func main() {
	fmt.Println(generateParenthesis(3))
}

File: lcode1-99/ex10/generateParenthesis.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        res = []
        # bfs加剪枝
        self.fn(res, "", 0, 0, n)
        return res

    def fn(self, res, string, left, right, n):
        if left > n or right > n or right > left:
            return
        # 递归的终止条件
        if left == n and right == n:
            res.append(string)
            return
        
        self.fn(res, string + "(", left+1, right, n)
        self.fn(res, string+")", left, right+1, n)
        return


sol = Solution()
print(sol.generateParenthesis(3))